src/hotstuff/protocol.rs

```diff
@@ -5,124 +5,550 @@
+                            proposal.block.hash, 
+                            vote_phase
+                        );
+
+                    if let HotStuffMessage::Vote(ref vote) = vote_msg {
+                        let vote_recipient = vote_recipient(&vote, &validator_set_state);
+                        self.sender_handle.send(vote_recipient, vote_msg.clone());
+                        block_tree.set_highest_view_voted(self.view_info.view)?;
+                        Event::Vote(VoteEvent{timestamp: SystemTime::now(), vote: vote.clone()}).publish(&self.event_publisher)
+                    }
+            }
+
+        }
+
+        // 6. Stop accepting proposals or nudges from this leader in this view.
+        match self.view_status {
+            ViewStatus::ViewInProgress => {
+                self.view_status = ViewStatus::LeaderProposed{leader: *origin}
+            },
+            ViewStatus::LeaderProposed{leader: _} => {
+                self.view_status = ViewStatus::ViewCompleted
+            },
+            _ => {}
+        }
+
+        Ok(())
     }
 
+    /// Process the received nudge.
     fn on_receive_nudge<K: KVStore>(
         &mut self, 
-        nudge: &Nudge, 
+        nudge: Nudge, 
         origin: &VerifyingKey,
         block_tree: &mut BlockTree<K>,
-    ) {
-        todo!()
+    ) -> Result<(), HotStuffError> {
+
+        Event::ReceiveNudge(ReceiveNudgeEvent{timestamp: SystemTime::now(), origin: *origin, nudge: nudge.clone()})
+        .publish(&self.event_publisher);
+
+        // 1. Check if the nudge is correct and safe.
+        if !nudge.justify.is_correct(block_tree)? || !safe_nudge(&nudge, self.view_info.view, block_tree, self.config.chain_id)? {
+            // Take note that proposals or nudges from this leader should no longer be accepted in this view.
+            match self.view_status {
+                ViewStatus::ViewInProgress => {
+                    self.view_status = ViewStatus::LeaderProposed{leader: *origin}
+                },
+                ViewStatus::LeaderProposed{leader: _} => {
+                    self.view_status = ViewStatus::ViewCompleted
+                },
+                _ => {}
+            }
+            return Ok(())
+        }
+
+        // 2. Trigger block tree updates: update highestQC, lock, commit.
+        let committed_validator_set_updates = 
+            update_block_tree(&nudge.justify, block_tree, &self.event_publisher)?;
+
+        if let Some(vs_updates) = committed_validator_set_updates {
+            self.validator_set_update_handle.update_validator_set(vs_updates)
+        }
+
+        // 3. Access the possibly updated validator set state, and update the vote collectors if needed.
+        let validator_set_state = block_tree.validator_set_state()?;
+
+        if self.vote_collectors.should_update_validator_sets(&validator_set_state) {
+            self.vote_collectors.update_validator_sets(&validator_set_state)
```

Comment:
In the same vein as my previous comment, I wonder whether or not it'd be better to move checking whether the validator set "should be updated" into `update_validator_sets` itself. If the provided validator set hasn't changed, then the function call will just quietly be a no-op.

Reply:
Logic moved to `update_validator_sets`.